testplan/testing/multitest/logging.py

```diff
@@ -0,0 +1,166 @@
+
+    def _attach_handler(
+        self,
+        result,
+        capture_level_override=None,
+        attach_log_override=None,
+        format_override=None,
+    ):
+        def override(value, _with):
+            return value if _with is None else _with
+
+        capture_level = override(
+            self.log_capture_config.capture_level, capture_level_override
+        )
+        save_to_file = override(
+            self.log_capture_config.attach_log, attach_log_override
+        )
+        format_string = override(
+            self.log_capture_config.format, format_override
+        )
+
+        if save_to_file:
+            stream = NamedTemporaryFile(
+                "w+t", dir=result._scratch, suffix=".log", delete=False
+            )
+        else:
+            stream = StringIO()
+
+        handler = StreamHandler(stream)
+        handler.setFormatter(logging.Formatter(format_string))
+
+        logger = self.select_logger(capture_level)
+
+        logger.addHandler(handler)
+
+        return self._LogCaptureInfo(
+            result, handler, save_to_file, capture_level
+        )
+
+    def _detach_handler(self, log_capture_info):
+        self.select_logger(log_capture_info.capture_level).removeHandler(
+            log_capture_info.handler
+        )
+        log_capture_info.handler.flush()
+        log_capture_info.handler.close()
+        if log_capture_info.attach_file:
+            log_capture_info.result.attach(
+                log_capture_info.handler.stream.name, CAPTURED_LOG_DESCRIPTION
+            )
+        else:
+            log_capture_info.result.log(
+                log_capture_info.handler.stream.getvalue(),
+                CAPTURED_LOG_DESCRIPTION,
+            )
+
+    @contextmanager
+    def capture_log(
+        self, result, capture_level=None, attach_log=None, format=None
+    ):
+        """Context manager to capture logs, capture the log in the provided result.
+
+        :param result: The result where to inject the log
+        :param CaptureLevel capture_level:  The level the log are captured, TESTSUITE (default), TESTPLAN or ROOT
+        :param bool attach_log: If True the logs captured to file and then attached to the result
+        :param str format: A format string can be passed to the loghandler
+        :return: returns the suite level logger
+        :rtype: logging.Logger
+        """
+        try:
+            info = self._attach_handler(
+                result,
+                capture_level_override=capture_level,
+                attach_log_override=attach_log,
+                format_override=format,
+            )
+
+            yield self.logger
+        finally:
+            self._detach_handler(info)
+
+    def select_logger(self, capture_level):
+        return capture_level(self)
+
+
+class AutoLogCaptureMixin(LogCaptureMixin):
+    def __init__(self):
+        super(AutoLogCaptureMixin, self).__init__()
+        self._state = None
+
```

Comment:
I found that there are some discrepancy in the current code about how pre_testcase shall be defined. One way is via @pre_testcase decorator - that's way in examples and tests. But then defining pre_testcase method in testsuite also kind of works, but in a slightly different way. We will need to consolidate this, but this shouldn't block your PR.